Ordnance Survey Geocoder provider

Adnan-cds opened the pull request:
Bringing in the Geocoder provider and its test file from localgovdrupal/localgov_forms.

Adnan-cds committed:
Config files and a working README.

Adnan-cds committed:
Our OS Places Geocoder provider.

Adnan-cds committed:
Unit test files.

Includes cached responses from the OS Places API.

Adnan-cds committed:
Github actions test runner.

First pass.

Adnan-cds committed:
Removing unneeded cached response file.

This one was added by mistake.

ekes commented on src/Model/LocalgovAddressInterface.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Adds UPRN and display value support to Location.
+ */
+interface LocalgovAddressInterface extends Location {
```
So this was where I was wondering if it wasn't better to make another interface (or more look at it) LocationUprn, LocationOsGrid. This means that any geocoder that returns UPRN could implement the interface too.

Scanning through Gecoder-PHP it's not what they're doing like they seem to make _final_ extended classes:
 https://github.com/geocoder-php/nominatim-provider/blob/master/Model/NominatimAddress.php but like in that example it _could_ implement a LocationUprn interface (but not a LocationOsGrid one).
 https://github.com/geocoder-php/open-cage-provider/blob/master/Model/OpenCageAddress.php - if the OSM one implemented an OSM interface for some OSM specific methods then the open cage address one could implement it!
Complete aside while I'm investigating this one it could also return the *Irish* Grid reference.
 https://github.com/geocoder-php/google-maps-provider/blob/master/Model/GoogleAddress.php

So I think this should either (a) plough the interface furrow as an example of what might be useful; or (b) just implement the _final_ class as Geocoder PHP existing practice.

Adnan-cds replied:
I have added two new interfaces:
- LocationUprnInterface
- LocationOsPlacesInterface which extends LocationUprnInterface

And two new Address record classes:
- UprnAddress
- OsPlacesAddress which extends UprnAddress

ekes replied:
Why do they each extend each other?

To my mind if it's worth making the interfaces (and not the final classes that geocoder-php usually does) it seems useful to allow the end consumer be able to query implements UprnInterface to see if it can pull a UPRN from the location that was returned to it; implements OsGridReference added by anything that gives UK OS Grid References; equally it could query something like implements OsmId (for all the multiple geocoders that return openstreetmap way/relation/node IDs). If they are specific to the geocoder itself, then the final class suffices?

Adnan-cds replied:
Sorry Ekes, I have to say I haven't understood everything you have said.  I am trying to explain as best as I can :(

> If they are specific to the geocoder itself, then the final class suffices?

The OsPlacesAddress class is specific to this geocoder.  That's why I have marked it as final.

The UprnAddress class (and the LocationUprnInterface interface) are generic and can be used by any geocoder wishing to deal with UPRNs.  This geocoder is probably not the best place to house these, but then I can't think of any other place without rolling a new package.  That's why I have just left them here.

ekes replied:
Easier to write in code than words I guess:

https://github.com/localgovdrupal/localgov_os_places_geocoder_provider/commit/f8da8f83c8b791ef33eaa89dc0ce31292c0fedc0

To have a OS Grid Reference doesn't need to have a UPRN (and less so, but still potentially true for a geocoder vice versa). So the interfaces don't need to inherit from each other. There's a bit more explanation in my example commit message.

Adnan-cds replied:
> To have a OS Grid Reference doesn't need to have a UPRN

I see the disconnect now.  I was under the impression that a UPRN is an indispensable part of an Ordnance Survey Places address, so wasn't considering any Address record with just a grid reference.  Anyway, [f8da8f8](https://github.com/localgovdrupal/localgov_os_places_geocoder_provider/commit/f8da8f83c8b791ef33eaa89dc0ce31292c0fedc0) looks good.  Should we merge the [interfaces](https://github.com/localgovdrupal/localgov_os_places_geocoder_provider/tree/interfaces) branch then?

ekes replied:
If you're good with that - and also can't think of an interface collection for the 'house name etc.' fields grouped together. Yep, let's go with it.

ekes commented on src/Model/LocalgovAddress.php:
```diff
@@ -0,0 +1,158 @@
+ *
+ * UPRN = Unique Property Reference Number.
+ */
+class LocalgovAddress extends Address implements LocalgovAddressInterface {
```
See note about interface, and _final_ on the interface.

It would also then be final class OsPlacesAddress I think, rather than Localgov.

Adnan-cds replied:
Done.

ekes commented on src/Provider/LocalgovOsPlacesGeocoder.php:
```diff
@@ -0,0 +1,166 @@
+ * @see https://osdatahub.os.uk/docs/places/overview
+ * @see https://en.wikipedia.org/wiki/Unique_Property_Reference_Number
+ */
+class LocalgovOsPlacesGeocoder extends AbstractHttpProvider implements GeocoderProviderInterface {
```
I'm guessing we should drop the Localgov part here too.

Adnan-cds replied:
Done.

Adnan-cds committed:
Support for Latitude and Longitude.

Search results now include Latitude and Longitude values.  Rich has pointed out
that requesting results in the "EPSG:4326" format returns latitude and longitude
values saving us from any conversion.

Adnan-cds committed:
Updated the test data to match latest results.

"7 Dyke road, Brighton" has disappeared from search results for
"Dyke road, Brighton"!  "11 Dyke road, Brighton" is the new first result item.

Search results also now include latitude and longitude values.

Adnan-cds committed:
Address record refactoring.

Adding two new address records in place of the previous LocalgovAddress record:
- A generic UPRN-based address record and its interface.
- An Ordnance Survey Places address record and its interface.

Adnan-cds committed:
Dropped Localgov prefix from Provider class name.

Adnan-cds committed:
Merge remote-tracking branch 'origin/1.x' into feature/provider

Adnan-cds commented:
Ready for review again.

finnlewis commented:
Hey @Adnan-cds - we're just discussing this in 'Merge Monday'. @ekes will try to find time to review soon. @j4-m might also be interested.

Adnan-cds commented:
Thanks Finn :)
